compiler/next/include/chpl/uast/Cobegin.h

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2021 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CHPL_UAST_COBEGIN_H
+#define CHPL_UAST_COBEGIN_H
+
+#include "chpl/queries/Location.h"
+#include "chpl/uast/BlockStyle.h"
+#include "chpl/uast/Expression.h"
+#include "chpl/uast/SimpleBlockLike.h"
+#include "chpl/uast/WithClause.h"
+
+namespace chpl {
+namespace uast {
+
+
+/**
+  This class represents a cobegin statement. For example:
+
+  \rst
+  .. code-block:: chapel
+
+      // Example 1:
+      var x = 0;
+      cobegin {
+        writeln(x);
+      }
+
+  \endrst
+
+ */
+class Cobegin final : public SimpleBlockLike {
```

Comment:
I'm not so sure it makes sense to talk about the individual things to run in parallel inside of a Cobegin as regular statements; rather they are tasks to create. However I understand it makes it easier to fit in.

Consider - if we were to make cobegin not inherit from SimpleBlockLike, all of the stmts in a SimpleBlockLike will run in order whenever they run. I think this is worth doing.

Reply:
I don't think we should make new accessors (e.g. `numTaskBodies()` `taskBody(int i)`) just for `Cobegin`. My position is that even if the statements in a `Cobegin` are lowered to become task bodies, syntactically they are just statements.

I understand the argument "if there is SimpleBlockLike, people will use it", and even given that people will use it, I do not think it is incorrect to view a `Cobegin` as containing a bunch of statements. Its body was parsed as a `block_stmt`. That they will be stuffed in a task bodies is a property of the parent `Cobegin` node, and not the statements themselves.

The `SimpleBlockLike` is intended to imply syntactic structure, not semantic structure. In the same way, the `IndexableLoop` is only implying syntactic structure. The lowered form of each indexable loop will be dramatically different depending on whether it `isFor()` or `isForall()`.

If you want to iterate over all the `SimpleBlockLike` things, already you are eliding semantics that you can only know exist by inspecting the fully derived class. And I think that if you cared about knowing that each statement in a Cobegin has to be lowered into its own task body, you could infer as much with `isCobegin()`.

Reply:
I agree with both sides of the argument. My preference is to have the Cobegin class reflect the semantics, i.e. have numTaskBodies/taskBody().
The argument in favor of Cobegin inheriting from SimpleBlockLike sounds to me like implementation expediency, which in this case does not carry much weight for me.

Reply:
It's true that there's a benefit of expediency, but I'd argue that from an interface perspective, it gives me no value to call `numTaskBodies()` or `taskBody()` in a completely immutable AST, because the thing I am accessing is at all points until lowering, still just a plain old statement. It serves as a reminder of what that statement will become after lowering, but that seems like a problem for the lowering pass and not for the user iterating over the block-like thing.

Reply:
Changed, see: e94d26b